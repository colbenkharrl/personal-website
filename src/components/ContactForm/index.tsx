'use client';

import { FC, useCallback } from 'react';
import { Button } from '../Button';
import { Input } from '../Input';
import { TextArea } from '../TextArea';
import { useRouter } from 'next/navigation';
// import { useFormStatus } from 'react-dom';

export type ContactFormProps = {
  onSubmit: (formData: FormData) => Promise<void>;
};

const ContactFormSubmitButton: FC = () => {
  // const { pending } = useFormStatus();

  return (
    <Button
      variant="primary"
      type="submit"
      className="w-full"
      // disabled={pending}
    >
      Submit
    </Button>
  );
};

export const ContactForm: FC<ContactFormProps> = ({ onSubmit }) => {
  const router = useRouter();

  const handleForm = useCallback(
    (formData: FormData) => {
      onSubmit(formData).then(() => router.push('/thank-you'));
    },
    [onSubmit, router],
  );

  return (
    // @ts-ignore - TS doesn't know about the FormData constructor
    <form action={handleForm} data-testid="ContactForm">
      <div className="space-y-12">
        <div className="border-b border-gray-900/10 pb-12">
          <div className="mt-10 grid grid-cols-1 gap-x-6 gap-y-8 sm:grid-cols-6">
            <Input
              containerProps={{ className: 'sm:col-span-3' }}
              type="text"
              label="First name"
              name="firstName"
              id="firstName"
              autoComplete="givenName"
              required
              invalidText="Please enter a valid first name."
              placeholder="Jane"
            />

            <Input
              containerProps={{ className: 'sm:col-span-3' }}
              type="text"
              label="Last name"
              name="lastName"
              id="lastName"
              autoComplete="familyName"
              required
              invalidText="Please enter a valid last name."
              placeholder="Doe"
            />

            <Input
              containerProps={{ className: 'col-span-full' }}
              type="email"
              label="Email address"
              name="email"
              id="email"
              autoComplete="email"
              required
              invalidText="Please enter a valid email address."
              placeholder="jdoe@email.com"
            />

            <TextArea
              containerProps={{ className: 'col-span-full' }}
              rows={3}
              label="Message"
              name="message"
              id="message"
              required
              invalidText="Please enter a message describing why you are contacting me."
              placeholder="I would like to chat about..."
            />
          </div>

          <div className="mt-6 flex items-center justify-start gap-x-6">
            <ContactFormSubmitButton />
          </div>
        </div>
      </div>
    </form>
  );
};
