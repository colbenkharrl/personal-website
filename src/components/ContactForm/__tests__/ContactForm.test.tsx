import { render, screen } from '@testing-library/react';
import * as stories from '../__stories__/ContactForm.stories';
import { composeStories } from '@storybook/react';

jest.mock('react-dom', () => ({
  ...jest.requireActual('react-dom'),
  useFormStatus: () => ({
    pending: false,
  }),
}));

const { Empty } = composeStories(stories);

it('ContactForm matches snapshot (regression test)', () => {
  render(<Empty />);
  expect(screen.getByTestId('ContactForm')).toMatchSnapshot(
    'Full component snapshot.',
  );
});
