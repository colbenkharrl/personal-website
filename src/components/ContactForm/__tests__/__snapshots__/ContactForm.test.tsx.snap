// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactForm matches snapshot (regression test): Full component snapshot. 1`] = `
<form
  data-testid="ContactForm"
>
  <div
    class="space-y-12"
  >
    <div
      class="border-b border-gray-900/10 pb-12"
    >
      <div
        class="mt-10 grid grid-cols-1 gap-x-6 gap-y-8 sm:grid-cols-6"
      >
        <div
          class="sm:col-span-3"
        >
          <label
            class="block text-sm font-medium leading-6 text-zinc-800 dark:text-zinc-200"
            for="firstName"
          >
            First name
            <span
              class="text-red-500"
            >
               *
            </span>
          </label>
          <div
            class="relative mt-2 rounded-md shadow-sm"
          >
            <input
              aria-invalid="false"
              autocomplete="givenName"
              class="block w-full rounded-md border-0 bg-transparent py-1.5 text-zinc-800 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-orange-500  dark:text-zinc-200 dark:focus:ring-orange-400 sm:text-sm sm:leading-6"
              id="firstName"
              name="firstName"
              placeholder="Jane"
              type="text"
            />
          </div>
        </div>
        <div
          class="sm:col-span-3"
        >
          <label
            class="block text-sm font-medium leading-6 text-zinc-800 dark:text-zinc-200"
            for="lastName"
          >
            Last name
            <span
              class="text-red-500"
            >
               *
            </span>
          </label>
          <div
            class="relative mt-2 rounded-md shadow-sm"
          >
            <input
              aria-invalid="false"
              autocomplete="familyName"
              class="block w-full rounded-md border-0 bg-transparent py-1.5 text-zinc-800 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-orange-500  dark:text-zinc-200 dark:focus:ring-orange-400 sm:text-sm sm:leading-6"
              id="lastName"
              name="lastName"
              placeholder="Doe"
              type="text"
            />
          </div>
        </div>
        <div
          class="col-span-full"
        >
          <label
            class="block text-sm font-medium leading-6 text-zinc-800 dark:text-zinc-200"
            for="email"
          >
            Email address
            <span
              class="text-red-500"
            >
               *
            </span>
          </label>
          <div
            class="relative mt-2 rounded-md shadow-sm"
          >
            <input
              aria-invalid="false"
              autocomplete="email"
              class="block w-full rounded-md border-0 bg-transparent py-1.5 text-zinc-800 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-orange-500  dark:text-zinc-200 dark:focus:ring-orange-400 sm:text-sm sm:leading-6"
              id="email"
              name="email"
              placeholder="jdoe@email.com"
              type="email"
            />
          </div>
        </div>
        <div
          class="col-span-full"
        >
          <label
            class="block text-sm font-medium leading-6 text-zinc-800 dark:text-zinc-200"
            for="message"
          >
            Message
            <span
              class="text-red-500"
            >
               *
            </span>
          </label>
          <div
            class="relative mt-2 rounded-md shadow-sm"
          >
            <textarea
              aria-invalid="false"
              class="block w-full rounded-md border-0 bg-transparent px-2 py-1.5 text-zinc-800 shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-400 focus:ring-2 focus:ring-inset focus:ring-orange-500 dark:text-zinc-200 dark:focus:ring-orange-400 sm:text-sm sm:leading-6"
              id="message"
              name="message"
              placeholder="I would like to chat about..."
              rows="3"
            />
          </div>
        </div>
      </div>
      <div
        class="mt-6 flex items-center justify-start gap-x-6"
      >
        <button
          class="disabled:opacity-30 inline-flex items-center gap-2 justify-center rounded-md py-2 px-3 text-sm outline-offset-2 transition active:transition-none bg-orange-500 font-semibold text-zinc-100 hover:bg-orange-600 active:bg-orange-500 active:text-zinc-100/70 dark:bg-orange-400 dark:hover:bg-orange-500 dark:active:bg-orange-700 dark:active:text-zinc-100/70 w-full"
          type="submit"
        >
          Submit
        </button>
      </div>
    </div>
  </div>
</form>
`;
